packages/mobile/src/identity/privacy.ts

```diff
@@ -0,0 +1,138 @@
+import { getPhoneHash, isE164Number } from '@celo/utils/src/phoneNumbers'
+import crypto from 'crypto'
+import { Platform } from 'react-native'
+import BlindThresholdBls from 'react-native-blind-threshold-bls'
+import { call, put, select } from 'redux-saga/effects'
+import { ErrorMessages } from 'src/app/ErrorMessages'
+import { PHONE_NUM_PRIVACY_SERVICE } from 'src/config'
+import { updateE164PhoneNumberSalts } from 'src/identity/actions'
+import { e164NumberToSaltSelector, E164NumberToSaltType } from 'src/identity/reducer'
+import Logger from 'src/utils/Logger'
+
+const TAG = 'identity/privacy'
+const SIGN_MESSAGE_ENDPOINT = '/getSalt' // TODO rename that endpoint
+
+export interface PhoneNumberHashDetails {
+  e164Number: string
+  phoneHash: string
+  salt: string
+}
+
+// Fetch and cache a phone number's salt and hash
+export function* fetchPhoneHashPrivate(e164Number: string) {
+  try {
+    Logger.debug(`${TAG}@fetchPrivatePhoneHash`, 'Fetching phone hash details')
+    const saltCache: E164NumberToSaltType = yield select(e164NumberToSaltSelector)
+    const cachedSalt = saltCache[e164Number]
+
+    if (cachedSalt) {
+      Logger.debug(`${TAG}@fetchPrivatePhoneHash`, 'Salt was cached')
+      const phoneHash = getPhoneHash(e164Number, cachedSalt)
+      return { e164Number, phoneHash, salt: cachedSalt }
+    }
+
+    Logger.debug(`${TAG}@fetchPrivatePhoneHash`, 'Salt was not cached, fetching')
+    const details: PhoneNumberHashDetails = yield call(getPhoneHashPrivate, e164Number)
+    yield put(updateE164PhoneNumberSalts({ [e164Number]: details.salt }))
+    return details
+  } catch (error) {
+    if (error.message === ErrorMessages.SALT_QUOTA_EXCEEDED) {
```

Comment:
should we be syncing on this error messages?

Reply:
As in syncing with what we do in the cloud service? No this is a wallet-specific thing. The place where the wallet + cloud need to be in sync is the error code (403)